CRM/Contribute/Tokens.php

```diff
@@ -73,35 +67,58 @@ protected function getExposedFields(): array {
       'thankyou_date',
       'tax_amount',
       'contribution_status_id',
+      'contribution_status_id:name',
+      'contribution_status_id:label',
       'financial_type_id',
+      'financial_type_id:name',
+      'financial_type_id:label',
       'payment_instrument_id',
+      'payment_instrument_id:name',
+      'payment_instrument_id:label',
       'cancel_reason',
       'amount_level',
       'check_number',
     ];
     if (CRM_Campaign_BAO_Campaign::isCampaignEnable()) {
       $fields[] = 'campaign_id';
+      $fields[] = 'campaign_id.name';
+      $fields[] = 'campaign_id.title';
     }
+
     return $fields;
   }
 
-  /**
-   * Get tokens supporting the syntax we are migrating to.
-   *
-   * In general these are tokens that were not previously supported
-   * so we can add them in the preferred way or that we have
-   * undertaken some, as yet to be written, db update.
-   *
-   * See https://lab.civicrm.org/dev/core/-/issues/2650
-   *
-   * @return string[]
-   */
-  public function getBasicTokens(): array {
-    $return = [];
-    foreach ($this->getExposedFields() as $fieldName) {
-      $return[$fieldName] = $this->getFieldMetadata()[$fieldName]['title'];
+  public function getAliasTokens(): array {
+    $aliases = [];
+    if (CRM_Campaign_BAO_Campaign::isCampaignEnable()) {
+      // Unit-tests are written to use these funny tokens - but they're not valid in APIv4.
+      $aliases['campaign'] = 'campaign_id.name';
+      $aliases['campaign_id:name'] = 'campaign_id.name';
+      $aliases['campaign_id:label'] = 'campaign_id.title';
+    }
+    return $aliases;
+  }
+
+  public function getPrefetchFields(\Civi\Token\Event\TokenValueEvent $e): array {
+    $result = parent::getPrefetchFields($e);
+
+    // Always prefetch 'civicrm_contribution.currency' in case we need to format other fields (fee_amount, total_amount, etc).
+    $result[] = 'currency';
+
+    return array_unique($result);
+  }
+
+  public function evaluateToken(\Civi\Token\TokenRow $row, $entity, $field, $prefetch = NULL) {
+    $values = $prefetch[$row->context[$this->getEntityIDField()]];
+
+    // Any monetary fields in a `Contribution` (`fee_amount`, `total_amount`, etc) should be formatted in matching `currency`.
+    // This formatting rule would be nonsensical in any other entity.
```

Comment:
Actually this formatting rule makes sense in contribution recur entity, event. For participant the field has a slightly different name & for membership we'd probably need to do something clever

 If we have an over-rideable getCurrency() it can fall back to the site default or potentially pick up the currency from the

Reply:
OK, so my reservation about putting `getCurrency()` into the base-class has been that there are multiple entities where it doesn't make sense (`contact`, `mailing`, `domain`, `actions`, `activity`,`case`). But I see your point that there are multiple entities for which it does make sense (`contribution`, `membership`, `contribution_recur`).

I suppose the base-class could have an implementation of `getCurrency()` which just throws an exception:

```php
public function getCurrency(string $field): string {
    throw new \CRM_Core_Exception("The entity (%s) does not support tokens with currency data. The entity must implement getCurrency().", $this->getEntityName());
}
```

Seems better than having a base method that makes a real decision (which would likely be the wrong decision - and give some false sense of security when extending to new currency-enabled entities), and it seems better than an `abstract` method (which would require several entities to add nonsensical implementations).

(The ontologically purest form might be some kind of `trait`, but I think that would bring its own bit of messiness that isn't worthwhile for one field.)

Reply:
hmm - I still prefer https://github.com/civicrm/civicrm-core/pull/21134/files#diff-5e5cdbc696384acab038bc1f37776a3450b0127b6a7606a326a49343399eee6cR384-R399 - with the fallback - which is more similar to what we do elsewhere